### ISSUE TITLE
Boolean indexing sometimes extremely slow

### ISSUE THREAD

**Issue by PeterKucirek** [pydata/pandas#11475]
I appear to have some code that is sometimes (inexplicably) running in pure Python (i.e. **not** Cython-optimized). Specifically, the problem is with `loc[]` indexing on a boolean Series:

![pandas slow loc lookup example](https://cloud.githubusercontent.com/assets/9703332/10828484/e9995bbc-7e4a-11e5-9b00-ae367f328541.png)

_(Apologies for the redacted code and console output, but this project is not an open-source delivery)_

For context: `world.persons` returns a `DataFrame` with approximately 8 million records long and 15 columns wide. It uses a MultiIndex with 2 levels.

This `run()` method is being called twice in the software sequence. The first time `run()` is called, `persons.loc[filter_column]` takes **15 minutes** to execute, while the second time it only takes 5 seconds.  What the heck is going on here?

Furthermore, this is NOT the first time in the program that `world.persons` is being indexed, so I am not sure how caching is being applied.

One other thing: prior to this `loc[]` call, my system starts to approach memory limits to the point that Windows begins paging.

My OS is Windows 7 x64, and I'm using Pandas version 0.17.0.

**Comment by jreback** [comment]
w/o more information this is impossible to even look at.

pls provide ``df.info()`` at the very least

further ``df.index.is_unique``

**Comment by PeterKucirek** [comment]
I do assert earlier in the code that `persons.index.is_unique` and that `persons.index.is_monotonic_increasing` and as far as I know that doesn't change. However I will provide a debug statement closer to the problem line.

**Comment by PeterKucirek** [comment]
Here's the result of `world.persons.info()`:
![persons info](https://cloud.githubusercontent.com/assets/9703332/10829034/2dfdc5ce-7e4e-11e5-9b64-db1c91124353.png)

**Comment by jreback** [comment]
```
In [24]: pd.options.display.max_rows=12

In [25]: N = 15000000

In [26]: df = DataFrame({'A' : np.random.randn(N)},index=pd.MultiIndex.from_product([np.arange(10000),np.arange(1500)],names=['first','second']))

In [27]: df.info()
<class 'pandas.core.frame.DataFrame'>
MultiIndex: 15000000 entries, (0, 0) to (9999, 1499)
Data columns (total 1 columns):
A    float64
dtypes: float64(1)
memory usage: 171.7+ MB

In [28]: df
Out[28]:
                     A
first second
0     0       0.579155
      1       0.087183
      2      -0.351394
      3      -1.583739
      4      -0.422606
      5       1.216556
...                ...
9999  1494    0.223166
      1495    0.438728
      1496    0.704697
      1497    0.164411
      1498    0.466204
      1499    1.556406

[15000000 rows x 1 columns]
In [29]: indexer = Series(np.random.randint(0,2,size=N).astype(bool),index=df.index)

In [30]: indexer
Out[30]:
first  second
0      0         False
       1          True
       2         False
       3          True
       4          True
       5         False
                 ...
9999   1494      False
       1495       True
       1496      False
       1497      False
       1498       True
       1499      False
dtype: bool

In [31]: df.loc[indexer]
Out[31]:
                     A
first second
0     1       0.087183
      3      -1.583739
      4      -0.422606
      7      -0.465112
      8      -0.035657
      9      -0.097889
...                ...
9999  1486    0.113131
      1487    3.654871
      1488    0.746310
      1492    0.713414
      1495    0.438728
      1498    0.466204

[7499448 rows x 1 columns]

In [32]: %timeit df.loc[indexer]
1 loops, best of 3: 237 ms per loop
```

**Comment by jreback** [comment]
if I had to guess, you are providing an indexer which is not ALIGNED with the index AT ALL, and it ends up creating a huge index.

so easiest thing to try is

``df.loc[indexer.values]`` which simply passes in a boolean numpy array which is not aligned (it must be the same length).

**Comment by PeterKucirek** [comment]
Is there a way I can test for such mis-alignment? 

Note that `filter_column` is coming out of `world.persons.eval("work_type == 'usual_workplace'")` so I'm confused as to how the result could have anything but the correct index. But I will try anyways.

**Comment by PeterKucirek** [comment]
I can confirm that the DataFrame is uniquely indexed and is monotonically increasing right before the problem point.

**Comment by jreback** [comment]
@PeterKucirek you seem to be subclassing DataFrame. I would never recommend you do that. Closing, but if you would like to provide additional details I could possibly help.

**Comment by PeterKucirek** [comment]
This is still a mystery to me. I re-ordered the program's execution so that this module runs one step earlier (the two steps are procedurally independent) and this indexing issue goes away, only to appear in the other module. 

I have indeed sub-classed DataFrame. The purpose behind doing so is to 'cache' relational indexing between associated tables. I could go into details, but the only relevant information is that I override `__finalize__` (but I am sure to `super()` it first) and make a call to `Index.get_indexer_for()`. So far I have not been able to determine whether this is the problem.

What bamboozles me is that nearly-identical calls are yielding different results and I cannot figure out why.

**Comment by jreback** [comment]
well there you go
Index.get_indexer_for() does not play well will duplicates (eg MultiIndexes)
further this is an internal API 

I'll reiterate subclassing is simply a bad idea
use container classes

**Comment by PeterKucirek** [comment]
Yes, it turns out that the problem was in `get_indexer_for`, although I note that the method just calls `get_indexer()` or `get_indexer_non_unique()` (which are part of the public API) accordingly.

I believe that what was happening was that an array of Python tuples was being created  and later collected by the GC when I began running out of memory. Thus the inconsistent behaviour.

I've re-implemented my `__finalize__` method to avoid that method entirely and the problem has been resolved.
